source/extensions/filters/common/ext_authz/ext_authz_impl.cc

```diff
@@ -47,23 +50,50 @@ void GrpcClientImpl::check(RequestCallbacks& callbacks,
 
 void GrpcClientImpl::onSuccess(
     std::unique_ptr<envoy::service::auth::v2alpha::CheckResponse>&& response, Tracing::Span& span) {
-  CheckStatus status = CheckStatus::OK;
   ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  ResponsePtr authz_response = std::make_unique<Response>(Response{});
+
   if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
-    status = CheckStatus::Denied;
     span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+    authz_response->status = CheckStatus::Denied;
+    authz_response->status_code = enumToInt(Http::Code::Forbidden);
+
+    if (response->has_denied_response()) {
+      for (const auto& header : response->denied_response().headers()) {
+        authz_response->headers.emplace_back(
+            std::make_pair(Http::LowerCaseString(header.first), header.second));
+      }
+      // Only reassign http status if it's greater than 0 and different than 403.
+      if (response->denied_response().status().code() &&
+          response->denied_response().status().code() != authz_response->status_code) {
```

Comment:
Nit: You can probably just drop the second conjunct here.

Reply:
👀